crates/ruma-client-api/src/error.rs

```diff
@@ -270,19 +291,123 @@ impl From<Error> for ErrorBody {
+    }
+}
+
+/// Errors in the `WWW-Authenticate` header.
+///
+/// To construct this use its `From<&str>` implementation. To get its serialized form, use its
+/// `Display` implementation.
+#[cfg(feature = "unstable-msc2967")]
+#[derive(Clone, Debug, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum AuthenticateError {
+    /// insufficient_scope
+    ///
+    /// Encountered when authentication is handled by OpenID Connect and the current access token
+    /// isn't authorized for the proper scope for this request. It should be paired with a
+    /// `401` status code and a `M_FORBIDDEN` error.
+    InsufficientScope {
+        /// The scope to request an authorization for.
+        scope: String,
+    },
+
+    #[doc(hidden)]
+    _Custom { errcode: PrivOwnedStr, attributes: AuthenticateAttrs },
+}
+
+#[cfg(feature = "unstable-msc2967")]
+#[doc(hidden)]
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct AuthenticateAttrs(BTreeMap<String, String>);
+
+#[cfg(feature = "unstable-msc2967")]
+impl AuthenticateError {
+    /// Construct an `AuthenticateError` from a string.
+    ///
+    /// Returns `None` if the string doesn't contain an error.
+    fn from_str(s: &str) -> Option<Self> {
+        if let Some(val) = s.strip_prefix("Bearer").map(str::trim) {
+            let mut errcode = None;
+            let mut attrs = BTreeMap::new();
+
+            // Split the attributes separated by commas and optionally spaces, then split the keys
+            // and the values, with the values optionally surrounded by double quotes.
+            for (key, value) in val
+                .split(',')
+                .filter_map(|attr| attr.trim().split_once('='))
+                .map(|(key, value)| (key, value.trim_matches('"')))
+            {
+                if key == "error" {
+                    errcode = Some(value);
+                } else {
+                    attrs.insert(key.to_owned(), value.to_owned());
+                }
+            }
+
+            if let Some(errcode) = errcode {
+                let error = if let Some(scope) =
+                    attrs.get("scope").filter(|_| errcode == "insufficient_scope")
+                {
+                    AuthenticateError::InsufficientScope { scope: scope.to_owned() }
+                } else {
+                    AuthenticateError::_Custom {
+                        errcode: PrivOwnedStr(errcode.into()),
+                        attributes: AuthenticateAttrs(attrs),
+                    }
+                };
+
+                return Some(error);
+            }
+        }
+
+        None
+    }
+}
+
+#[cfg(feature = "unstable-msc2967")]
+impl fmt::Display for AuthenticateError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            AuthenticateError::InsufficientScope { scope } => {
+                f.write_str(&format!("Bearer error=\"insufficient_scope\", scope=\"{scope}\""))
```

Comment:
```suggestion
                write!(f, "Bearer error=\"insufficient_scope\", scope=\"{scope}\"")
```

same below (if we continue using `fmt` stuff for this)